From b61b0b0dfd28cc3b196b81e0a223519199b71249 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Fri, 10 Dec 2021 16:12:06 +0800
Subject: [PATCH 53/81] modesetting: Delay link failure re-set when DPMS not on

Some monitors might report bad link right after DPMS turned off, delay
the link failure mode re-set to avoid bringing them back on.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 hw/xfree86/drivers/modesetting/drmmode_display.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/hw/xfree86/drivers/modesetting/drmmode_display.c b/hw/xfree86/drivers/modesetting/drmmode_display.c
index f6cacdbbf..9c1e67752 100644
--- a/hw/xfree86/drivers/modesetting/drmmode_display.c
+++ b/hw/xfree86/drivers/modesetting/drmmode_display.c
@@ -4305,6 +4305,13 @@ drmmode_update_kms_state(drmmode_ptr drmmode)
 
             if (!strcmp(p->mode_prop->name, "link-status")) {
                 if (p->value == DRM_MODE_LINK_STATUS_BAD) {
+                    drmmode_crtc_private_ptr drmmode_crtc = crtc->driver_private;
+                    drmmode_crtc->need_modeset = TRUE;
+
+                    /* Delay re-set to dpms-on */
+                    if (drmmode_output->dpms != DPMSModeOn)
+                        break;
+
                     /* the connector got a link failure, re-set the current mode */
                     drmmode_set_mode_major(crtc, &crtc->mode, crtc->rotation,
                                            crtc->x, crtc->y);
-- 
2.38.5

